services/federator/src/Federator/Monitor/Internal.hs

```diff
@@ -0,0 +1,355 @@
+          . Log.field "descriptor" (show w)
+          . Log.field "path" pathText
+    Left e -> do
+      Log.err $
+        Log.msg ("error while try to add file watch" :: Text)
+          . Log.field "path" pathText
+          . Log.field "error" (displayException e)
+
+addWatchAndSave ::
+  INotify ->
+  [EventVariety] ->
+  IORef Watches ->
+  WatchedPath ->
+  (Event -> IO ()) ->
+  IO WatchDescriptor
+addWatchAndSave inotify events watchesVar wpath handler = do
+  let path = watchedPath wpath
+  -- create a new watch
+  w' <- addWatch inotify events path handler
+  -- atomically save it in the map, and return the old one
+  mw <-
+    atomicModifyIORef watchesVar $
+      swap . Map.alterF (,Just (w', wpath)) path
+  -- remove the old watch
+  case mw of
+    Nothing -> pure ()
+    Just (w, _) -> removeWatch w
+  pure w'
+
+certificatePaths :: RunSettings -> [FilePath]
+certificatePaths rs =
+  maybeToList (remoteCAStore rs)
+    ++ [ clientCertificate rs,
+         clientPrivateKey rs
+       ]
+
+certificateWatchPaths :: RunSettings -> IO (Set WatchedPath)
+certificateWatchPaths =
+  fmap (mergePaths . concat)
+    . traverse watched
+    . certificatePaths
+  where
+    watched :: FilePath -> IO [WatchedPath]
+    watched path = do
+      rpath <- rawPath path
+      dirs <- watchedDirs path
+      pure $ WatchedFile rpath : dirs
+
+    watchedDirs :: FilePath -> IO [WatchedPath]
+    watchedDirs path = do
+      let (dir, base) = splitFileName (dropTrailingPathSeparator path)
+      if dir == path
+        then pure [] -- base case: root directory
+        else do
+          wds <- watchedDirs dir
+          rdir <- rawPath (dropTrailingPathSeparator dir)
+          rbase <- rawPath base
+          pure $ WatchedDir rdir (Set.singleton rbase) : wds
+
+data FederationSetupError
+  = InvalidCAStore FilePath
+  | InvalidClientCertificate String
+  deriving (Show)
+
+instance Exception FederationSetupError
+
+showFederationSetupError :: FederationSetupError -> Text
+showFederationSetupError (InvalidCAStore path) = "invalid CA store: " <> Text.pack path
+showFederationSetupError (InvalidClientCertificate msg) = Text.pack msg
+
+mkTLSSettings ::
+  Members '[Embed IO, Polysemy.Error FederationSetupError] r =>
+  RunSettings ->
+  Sem r TLSSettings
+mkTLSSettings settings =
+  TLSSettings
+    <$> mkCAStore settings
+    <*> mkCreds settings
+
+mkCAStore ::
+  Members '[Embed IO, Polysemy.Error FederationSetupError] r =>
+  RunSettings ->
+  Sem r CertificateStore
+mkCAStore settings = do
+  customCAStore <- fmap (fromRight mempty) . Polysemy.runError @() $ do
+    path <- maybe (Polysemy.throw ()) pure $ remoteCAStore settings
+    embed (readCertificateStore path)
+      >>= maybe (Polysemy.throw (InvalidCAStore path)) pure
+  systemCAStore <-
+    if useSystemCAStore settings
+      then embed getSystemCertificateStore
+      else pure mempty
+  pure (customCAStore <> systemCAStore)
+
+mkCreds ::
+  forall r.
```

Comment:
Just nit-picking: perhaps this quantification is not used anywhere?

Reply:
That's right, thanks. Fixed in 0449a7904f1b20726d2cc0bc228c41cfc5d19b78.